Sources/StreamChat/Controllers/BackgroundListDatabaseObserver.swift

```diff
@@ -0,0 +1,297 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+class ListDatabaseObserverWrapper<Item, DTO: NSManagedObject> {
+    private var foreground: ListDatabaseObserver<Item, DTO>?
+    private var background: BackgroundListDatabaseObserver<Item, DTO>?
+    let isBackground: Bool
+
+    var items: LazyCachedMapCollection<Item> {
+        if isBackground, let background = background {
+            return background.items
+        } else if let foreground = foreground {
+            return foreground.items
+        } else {
+            log.assertionFailure("Should have foreground or background observer")
+            return []
+        }
+    }
+
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerWillChangeContent`
+    /// on its delegate.
+    var onWillChange: (() -> Void)? {
+        didSet {
+            if isBackground {
+                background?.onWillChange = onWillChange
+            } else {
+                foreground?.onWillChange = onWillChange
+            }
+        }
+    }
+
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerDidChangeContent`
+    /// on its delegate.
+    var onDidChange: (([ListChange<Item>]) -> Void)? {
+        didSet {
+            if isBackground {
+                background?.onDidChange = onDidChange
+            } else {
+                foreground?.onChange = onDidChange
+            }
+        }
+    }
+
+    init(
+        isBackground: Bool,
+        database: DatabaseContainer,
+        fetchRequest: NSFetchRequest<DTO>,
+        itemCreator: @escaping (DTO) throws -> Item,
+        fetchedResultsControllerType: NSFetchedResultsController<DTO>.Type = NSFetchedResultsController<DTO>.self
+    ) {
+        self.isBackground = isBackground
+        if isBackground {
+            background = BackgroundListDatabaseObserver(
+                context: database.backgroundReadOnlyContext,
+                fetchRequest: fetchRequest,
+                itemCreator: itemCreator,
+                fetchedResultsControllerType: fetchedResultsControllerType
+            )
+        } else {
+            foreground = ListDatabaseObserver(
+                context: database.viewContext,
+                fetchRequest: fetchRequest,
+                itemCreator: itemCreator,
+                fetchedResultsControllerType: fetchedResultsControllerType
+            )
+        }
```

Comment:
Another option, is to maintain the Wrapper, but here, we only have one observer, that is initialised depending on isBackground. In that case, the if, is only here, and not spread across the wrapper